[Merged by Bors] - feat(LinearAlgebra/Projectivization/Subspace): correspondence between linear and projective subspaces

gasparattila opened the pull request:
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

gasparattila committed:
feat(LinearAlgebra/Projectivization/Subspace): correspondence between linear and projective subspaces

eric-wieser commented on Mathlib/LinearAlgebra/Projectivization/Subspace.lean:
```diff
@@ -207,3 +205,53 @@ theorem span_eq_span_iff {S T : Set (ℙ K V)} : span S = span T ↔ S ⊆ span
+    }
+    invFun s := {
+      carrier := {x | (h : x ≠ 0) → Projectivization.mk K x h ∈ s.carrier}
+      add_mem' := by
+        intro x y hx₁ hy₁
```
```suggestion
      add_mem' {x y} hx₁ hy₁ := by
```

eric-wieser commented on Mathlib/LinearAlgebra/Projectivization/Subspace.lean:
```diff
@@ -207,3 +205,53 @@ theorem span_eq_span_iff {S T : Set (ℙ K V)} : span S = span T ↔ S ⊆ span
+  (fun _ _ h₁ _ h₂ _ => h₁ <| h₂ _)
+
+@[simp]
+theorem mem_projectivization_iff (s : Submodule K V) {v : V} (hv : v ≠ 0) :
```
Can you add a similar lemma for `projectivization.symm s` too?

eric-wieser commented on Mathlib/LinearAlgebra/Projectivization/Subspace.lean:
```diff
@@ -207,3 +205,53 @@ theorem span_eq_span_iff {S T : Set (ℙ K V)} : span S = span T ↔ S ⊆ span
+
+variable {K V}
+
+/-- Subspaces of a vector space correspond to subspaces of its projectivization. -/
```
Since this is a `def`, you should describe _what_ the correspondence is.

gasparattila committed:
Rework

gasparattila commented:
Thanks for the suggestions! I realized that the inverse is inconvenient to use, so I added an abbrev for the inverse. I also added a lemma relating the definition to the submodule of a single point.

Vierkantor reviewed (commented):
I have a few comments, overall it looks quite nice. Do you have any further work building on this, so I can get a bit of context?

gasparattila committed:
Update Mathlib/LinearAlgebra/Projectivization/Subspace.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

gasparattila commented:
Thanks for the suggestions! The motivation for this is defining bundled projective transforms. This could be done using a class like this:
```lean
class IsProjectivization K V P [Field K] [Module K V] where
  equiv : P ≃ Projectivization K V
```
Then we would have an instance `IsProjectivization K s.submodule s` for `s : Projectivization.Subspace K P` to allow defining maps between subspaces. But I'm not sure if this is the best design and I'm currently not planning to work on this.

-awaiting-author

Vierkantor commented on Mathlib/LinearAlgebra/Projectivization/Subspace.lean:
```diff
@@ -204,6 +202,70 @@ theorem span_eq_span_iff {S T : Set (ℙ K V)} : span S = span T ↔ S ⊆ span
+    mem_add' _ _ _ _ _ h₁ h₂ := s.add_mem h₁ h₂ }
+  left_inv s := by
+    ext x
+    cases x with | _ x hx =>
```
`ext` can also do case distinctions!
```suggestion
    ext ⟨x, hx⟩
```

Vierkantor reviewed (approved):
I took a look at the proofs and found some tricks to make them a bit shorter and more elegant. With those changes, we should be ready to merge, thank you for all the work! 🎉 

bors d+

gasparattila committed:
Apply suggestions from code review

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

gasparattila commented:
bors r+
